### ISSUE TITLE
Missing text rendering support

### ISSUE THREAD

**Issue by tymlipari** [BabylonJS/BabylonNative#616]
Currently Babylon Native doesn't render text for controls in Babylon.GUI.

**Comment by bghgary** [comment]
Related #423

**Comment by CedricGuillemet** [comment]
### A bit of explanation why the issue is renamed from 'text rendering' to 'canvas polyfill':

Many rendered elements in bjs are done through HTML element. The HTML element surface is then copied into a texture and used by bjs. It's the case for GUI, webcam, video,...

Example with drawtext in dynamictexture.ts:
- texture and canvas are created in the constructor

https://github.com/BabylonJS/Babylon.js/blob/aa8ac1ca1bbcff3754fe1df74ab075a678afc957/src/Materials/Textures/dynamicTexture.ts#L46

- draw text using the canvas context

https://github.com/BabylonJS/Babylon.js/blob/aa8ac1ca1bbcff3754fe1df74ab075a678afc957/src/Materials/Textures/dynamicTexture.ts#L169

- texture is updated

https://github.com/BabylonJS/Babylon.js/blob/aa8ac1ca1bbcff3754fe1df74ab075a678afc957/src/Materials/Textures/dynamicTexture.ts#L137

Text rendering in whole bjs (except 3d text mesh) is done the same way. In order to be consistent with the engine, we must do text rendering (and 2D shape rendering) with an abstraction of the Canvas.

### Canvas polyfill PoC
A draft PR is available here : https://github.com/BabylonJS/BabylonNative/pull/441
It creates 2 main new JS objects: `OffscreenCanvas` and `Context`

OffscreenCanvas is a wrapping around a Render Target.
Context provides functions to draw inside the OffscreenCanvas (rectangle, rotate, ...)

### What's left to do in the PR ?
- A proper abstraction for the GUI (like NativeEngine interface). Provide an object (Canvas on the web, NativeCanvas for Native) : 3-5 days of work. Mostly on babylonjs repo
- RenderTarget management. Each new OffscreenCanvas creates a new rendertarget with bgfx. ViewId and life cycle is not taking NativeEngine into account. Moreover, more RenderTarget than supported by bgfx can be created. This part needs Gary's PR with new RenderTarget management. 5 days of work
- Code missing Canvas functions. I used to code them when missing while trying to run some PG. Once the abstraction is done, an exhaustive list of functions to support will be available. 3 days of work
- Font loading. In the browser, the font is loaded and available when you render text Here, we need to load a .ttf. On Windows, we can get the system path and load what's required. On Apple, you can load a font without using the UIKit. On Android, you are on your own. I need inputs and use cases here to drive the development: provide 'stand fonts' in the playground or plugin. Let the user download/pacakge them, ... 3 days of work
- Font rendering with multisize support. Only 1 text size supported for now. Add support for variable size. 3 days of work.
- Miscellaneous fixes when testing on other platforms. 5 days of work.

Total : 22-24 days of work.

**Comment by CedricGuillemet** [comment]
- [ ] Canvas abstraction. (@CedricGuillemet )
To make it clear that we use a subset of Canvas/Context in the engine, all the calls on these objects are done through an interface (ICanvas.ts). https://github.com/BabylonJS/Babylon.js/pull/10043 A build of this branch is pushed in this BabylonNative PR. 
- [ ] Missing functions in NativeEngine.ts
https://github.com/CedricGuillemet/Babylon.js/blob/d4cd8b3ecde64af3bbd9a991bf0e1299ef17e524/src/Engines/nativeEngine.ts#L2354 is done implemented for native.
- [ ] DrawImage implementation
Particular attention to all context function using images. Some partial work has been done https://github.com/CedricGuillemet/Babylon.js/blob/d4cd8b3ecde64af3bbd9a991bf0e1299ef17e524/src/Engines/nativeEngine.ts#L1827 Use the same objects when implementing DrawImage or modify that one. This function is used to create ground mesh with height map is is tested by the validation tests.
- [ ] Object life cyle
Objects (canvas + context, possibily other) are not properly disposed when the JS engine is destroyed. This can be checked in Debug build with the Playground app. When exiting the app bgfx dumps resource leaks. Make sure memory/resource leaks are fixed. (debug build + debug run)
- [ ] Missing RenderingContext functions.
This interface : https://github.com/CedricGuillemet/Babylon.js/blob/CanvasAbstraction/src/Engines/ICanvas.ts
lists all the function used by Babylonjs. Some are implemented, some are not. I suggest to create a playground that tests every function, then make sure it also works with BabylonNative Playground App. This PG should be used in the validation tests.
Nanovg should provide all the API needed to make it work.  Most of this code will be implemented in `Context.cpp`. 
- [ ] Render target management. 
Each Canvas has an associated render target https://github.com/CedricGuillemet/BabylonNative/blob/dcec1cd0635e57664f071b3f30316df6dc661a23/Polyfills/Canvas/Source/Canvas.cpp#L95 Because of bgfx limited number of render targets, the number of active canvas is also limited. A more efficient way would be to create texture but only use a render target when the context is dirty. https://github.com/CedricGuillemet/BabylonNative/blob/dcec1cd0635e57664f071b3f30316df6dc661a23/Polyfills/Canvas/Source/Context.cpp#L485
- [ ] Test and debug on iOS, Android. 
Also test Linux et MacOS but if iOS and Android are working, I'm confident it would need much more. Use Validation tests with PG mentioned earlier or the Playground App. Same UWP is using D3D11 so once Win32 version is ok, UWP should also be ok but needs tests to validate.

**Comment by chrisfromwork** [comment]
@CedricGuillemet has this work thus far been primarily targetted at 3d gui controls? I setup a local test environment to setup 2d controls and am seeing native exceptions when calling AdvancedDynamicTexture.CreateFullscreenUI. Is this a known/documented issue?

**Comment by chrisfromwork** [comment]
@CedricGuillemet should this work be used to track the fact that layers don't work in babylonnative? or should that be tracked with a separate issue?
